crates/provider/src/quorum/transport.rs

```diff
@@ -0,0 +1,240 @@
+                + Send
+                + 'static,
+        >,
+    >;
+
+    fn poll_ready(
+        &mut self,
+        _cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        std::task::Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, req: RequestPacket) -> Self::Future {
+        let requests = self
+            .transports
+            .iter()
+            .enumerate()
+            .map(|(id, transport)| PendingRequest {
+                future: transport.inner.clone().call(req.clone()),
+                id,
+            })
+            .collect::<Vec<_>>();
+
+        let quorum_request = QuorumRequest::new(self.clone(), requests);
+
+        Box::pin(quorum_request)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct WeightedTransport {
+    inner: BoxTransport,
+    pub weight: u64,
+}
+
+impl WeightedTransport {
+    pub fn new(inner: BoxTransport) -> Self {
+        Self { inner, weight: 1 }
+    }
+
+    pub fn with_weight(inner: BoxTransport, weight: u64) -> Self {
+        Self { inner, weight }
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct QuorumTransportBuilder {
+    quorum: Quorum,
+    transports: Vec<WeightedTransport>,
+}
+
+impl QuorumTransportBuilder {
+    pub fn add_transports(
+        mut self,
+        providers: impl IntoIterator<Item = WeightedTransport>,
+    ) -> Self {
+        for provider in providers {
+            self.transports.push(provider);
+        }
+        self
+    }
+
+    /// Set the kind of quorum
+    pub fn quorum(mut self, quorum: Quorum) -> Self {
+        self.quorum = quorum;
+        self
+    }
+
+    pub fn build(self) -> QuorumTransport {
+        let quorum_weight = self.quorum.weight(&self.transports);
+        QuorumTransport {
+            transports: self.transports,
+            quorum_weight,
+        }
+    }
+}
+
+pub struct QuorumRequest<'a> {
+    inner: QuorumTransport,
+    /// The different answers with their cumulative weight
+    responses: Vec<(ResponsePacket, u64)>,
+    /// All the errors the provider yielded
+    errors: Vec<TransportError>,
+    // Requests currently pending
+    requests: Vec<PendingRequest<'a>>,
+}
+
+type PendingRequestFuture<'a> = Pin<
+    Box<
+        dyn Future<Output = Result<ResponsePacket, RpcError<TransportErrorKind>>>
+            + 'a
+            + Send,
+    >,
+>;
+
+struct PendingRequest<'a> {
+    future: PendingRequestFuture<'a>,
+    id: usize,
+}
+
+impl Future for PendingRequest<'_> {
+    type Output = Result<ResponsePacket, RpcError<TransportErrorKind>>;
+    fn poll(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Self::Output> {
+        self.future.as_mut().poll(cx)
+    }
+}
+
+impl<'a> QuorumRequest<'a> {
+    fn new(inner: QuorumTransport, requests: Vec<PendingRequest<'a>>) -> Self {
+        Self {
+            responses: Vec::new(),
+            errors: Vec::new(),
```

Comment:
this can be initialized with the same capacity as the number of providers


```suggestion
            errors: Vec::with_capacity(...),
```